source/server/filter_chain_manager_impl.h

```diff
@@ -134,26 +136,54 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
   const Network::Address::InstanceConstSharedPtr address_;
 };
 
-class FilterChainImpl : public Network::FilterChain {
+/**
+ * Base implementation for TCP and QUIC filter chain.
+ */
+class FilterChainImplBase : public Network::FilterChain {
 public:
-  FilterChainImpl(Network::TransportSocketFactoryPtr&& transport_socket_factory,
-                  std::vector<Network::FilterFactoryCb>&& filters_factory)
-      : transport_socket_factory_(std::move(transport_socket_factory)),
-        filters_factory_(std::move(filters_factory)) {}
+  FilterChainImplBase(std::vector<Network::FilterFactoryCb>&& filters_factory)
+      : filters_factory_(std::move(filters_factory)) {}
 
   // Network::FilterChain
-  const Network::TransportSocketFactory& transportSocketFactory() const override {
-    return *transport_socket_factory_;
-  }
-
   const std::vector<Network::FilterFactoryCb>& networkFilterFactories() const override {
     return filters_factory_;
   }
 
 private:
-  const Network::TransportSocketFactoryPtr transport_socket_factory_;
   const std::vector<Network::FilterFactoryCb> filters_factory_;
 };
 
+class TcpFilterChainImpl : public FilterChainImplBase {
+public:
+  TcpFilterChainImpl(Network::TransportSocketFactoryPtr&& transport_socket_factory,
+                     std::vector<Network::FilterFactoryCb>&& filters_factory)
+      : FilterChainImplBase(std::move(filters_factory)),
+        transport_socket_factory_(std::move(transport_socket_factory)) {}
+
+  // Network::FilterChain
+  const Network::TransportSocketFactory* transportSocketFactory() const override {
+    return transport_socket_factory_.get();
+  }
+
+private:
+  const Network::TransportSocketFactoryPtr transport_socket_factory_;
+};
+
+class QuicFilterChainImpl : public FilterChainImplBase {
+public:
+  QuicFilterChainImpl(std::unique_ptr<Ssl::ContextConfig> tls_context_config,
+                      std::vector<Network::FilterFactoryCb>&& filters_factory)
+      : FilterChainImplBase(std::move(filters_factory)),
+        tls_context_config_(std::move(tls_context_config)) {}
+
+  // Network::FilterChain
+  const Network::TransportSocketFactory* transportSocketFactory() const override { return nullptr; }
```

Comment:
It seems you're making trasnportSocketFactory nullable and then introduce direct TLS context config here, what is the rationale behind this? This doesn't look quite right as TLS is a part of extension and the abstraction is desired. Can't a QUIC specific transport socket factory (and transport socket) does this instead? That allows the transport socket be wrapped by another (e.g. TAP).

Also tagging #8098, I'm not thrilled adding more direct dependency of tls_context, but in favor of wrapping it in an opaque config of transport socket.

Reply:
I see. It seems that the spirit of this PR goes the opposite of #8098. Actually in the design doc, I even proposed to move tls_context outside of transport socket. The rationale behind is that QUIC doesn't have transport socket at all. Only TCP has the concept of listening socket and transport socket, but QUIC only has one socket.

Reply:
The transport socket is an Envoy concept in config and implementation, it's not protocol semantics either. From config point it seems possible to make the config looks like:
```
filter_chain_match:
transport_socket:
  name: quic
  config: 
    {tls_context}
filters:
```

At implementation level, ContextConfigFactory and ContextConfig proposed in this PR are at same layer of TransportSocketConfigFactory and TransportSocketFactory respectively, but more specific to TLS, which the latter tried to abstract from. I think what they are trying to do are pretty similar so should be achievable by modifying TransportSocketConfigFactory and TransportSocketFactory to match what is needed for QUIC.

@mattklein123 @alyssawilk WDYT?

Reply:
The whole point of having quic transport socket is to be able to access tls_context in quic implemention of TransportSocketFactory then. Is it an overkill? What should the socket implementation actually look like?

Reply:
Generally +1 to what @lizan is saying. I think this is the crux of my other comments in this PR but I don't know the code as well as @lizan. Something generally seems a bit off here and it seems like we should be able to simplify all of this somehow.

I'm not sure there needs to be an actual socket in implementation, but mostly around configuration and providing the TLS context config?

@lizan is it possible for you to sync briefly with @danzh2010 on this? I could dig into the code more myself but I will need to spend more time paging it in.

Reply:
> The whole point of having quic transport socket is to be able to access tls_context in quic implemention of TransportSocketFactory then. Is it an overkill? What should the socket implementation actually look like?

I think you only need QuicTransportSocketFactory for now. It shouldn't be overkill if you make createTransportSocket returns nullptr, and provide an accessor of SslContextConfig. The `TransportSocketFactory` interface only has two methods.

> I'm not sure there needs to be an actual socket in implementation, but mostly around configuration and providing the TLS context config?

Based on the discussion with @danzh2010 today, probably it is not quite possible to make a transport socket implementation with current L7/L4 split situation today. While from my (very rough) code reading of quiche, the possible transport socket implementation looks like combination of `quic::ProcessPacketInterface`, `quic::QuicPacketWriter` and connection_id handling (as IoHandle).

Reply:
I reverted QuicFilterChain and added QuicTransportSocketFactory as you suggested. PTAL!

Reply:
This part LGTM